tests/tlsio_esp8266_ut/tlsio_esp8266_ut.c

```diff
@@ -0,0 +1,733 @@
+        TLS_IO_INSTANCE instance;
+        instance.tlsio_state = TLSIO_STATE_OPEN;
+        g_ssl_write_success = 0;
+
+        const IO_INTERFACE_DESCRIPTION* tlsioInterfaces = tlsio_openssl_get_interface_description();
+        ASSERT_IS_NOT_NULL(tlsioInterfaces);
+
+        ///act
+        result = tlsioInterfaces->concrete_io_send(&instance, NULL, 10, NULL, NULL);
+        
+        ///assert
+        ASSERT_ARE_NOT_EQUAL(int, result, 0);
+        ///cleanup
+    }
+
+    /* Tests_SRS_TLSIO_SSL_ESP8266_99_015: [ The tlsio_openssl_send wrong state.]*/
+    TEST_FUNCTION(tlsio_openssl_send_wrong_state__failed)
+    {
+        ///arrange
+        int result;
+        const IO_INTERFACE_DESCRIPTION* tlsioInterfaces = tlsio_openssl_get_interface_description();
+        ASSERT_IS_NOT_NULL(tlsioInterfaces);
+
+        TLS_IO_INSTANCE instance;
+        instance.tlsio_state = TLSIO_STATE_NOT_OPEN;
+        ///act
+        result = tlsioInterfaces->concrete_io_send(&instance, NULL, 0,  NULL, NULL);
+
+        ///assert
+        ASSERT_ARE_NOT_EQUAL(int, 0, result);
+
+        ///cleanup
+    }
+
+
+    /* Tests_SRS_TLSIO_SSL_ESP8266_99_014: [ The tlsio_openssl_send NULL instance.]*/
+    TEST_FUNCTION(tlsio_openssl_send__failed)
+    {
+        ///arrange
+        int result;
+        const IO_INTERFACE_DESCRIPTION* tlsioInterfaces = tlsio_openssl_get_interface_description();
+        ASSERT_IS_NOT_NULL(tlsioInterfaces);
+
+        ///act
+        result = tlsioInterfaces->concrete_io_send(NULL, NULL, 0, NULL, NULL);
+
+        ///assert
+        ASSERT_ARE_NOT_EQUAL(int, 0, result);
+
+        ///cleanup
+    }
+
+    /* Tests_SRS_TLSIO_SSL_ESP8266_99_013: [ The tlsio_openssl_close succeed.]*/
+    TEST_FUNCTION(tlsio_openssl_close__succeed)
+    {
+        ///arrange
+        TLS_IO_INSTANCE instance;
+        instance.tlsio_state = TLSIO_STATE_OPEN;
+
+        const IO_INTERFACE_DESCRIPTION* tlsioInterfaces = tlsio_openssl_get_interface_description();
+        ASSERT_IS_NOT_NULL(tlsioInterfaces);
+
+        umock_c_reset_all_calls();
+        STRICT_EXPECTED_CALL(SSL_free(IGNORED_PTR_ARG)).IgnoreArgument(1);
+        STRICT_EXPECTED_CALL(SSL_CTX_free(IGNORED_PTR_ARG)).IgnoreArgument(1);
+
+        ///act
+        tlsioInterfaces->concrete_io_close(&instance, NULL, NULL);
+        
+        ///assert
+        ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+        ///cleanup
+    }
+
+
+    /* Tests_SRS_TLSIO_SSL_ESP8266_99_012: [ The tlsio_openssl_close wrong state.]*/
+    TEST_FUNCTION(tlsio_openssl_close_wrong_state__failed)
+    {
+        ///arrange
+        int result;
+        const IO_INTERFACE_DESCRIPTION* tlsioInterfaces = tlsio_openssl_get_interface_description();
+        ASSERT_IS_NOT_NULL(tlsioInterfaces);
+
+        TLS_IO_INSTANCE instance;
+        instance.tlsio_state = TLSIO_STATE_NOT_OPEN;
+        ///act
+        result = tlsioInterfaces->concrete_io_close(&instance, NULL, NULL);
```

Comment:
Done. Added checks for `TLSIO_STATE_ERROR` and `TLSIO_STATE_NOT_OPEN` for final state.